### ISSUE TITLE
Request : optimize loading time

### ISSUE THREAD

**Issue by ChameleonScales** [PhieF/CarnetNextcloud#73]
I really like your project and use it a lot.
However, I have tried it on several nextcloud servers and one thing becomes clear : there is a major loading time issue.
Opening notes takes disproportionally long and this can quickly become a time sink when you navigate between notes frequently.
There's plenty to do to make this app even more awesome but I think loading optimization should be addressed soon to set a solid base.

**Comment by PhieF** [comment]
I'm not sure I can do better and I know first loading time can be pretty long. about 3 secs on first and 1sec on second on my server, how long does it take on yours ?

**Comment by ChameleonScales** [comment]
9 sec for the first open and then it ranges between 6 and 8 sec.
The notes only contain a few small lines of text without any media.
Meanwhile, opening a text file from the file manager app takes less than a second and opening a pdf with a big image in it less than 2 (on the same server).

**Comment by auberginepop** [comment]
I am running it on Nextcloud in a snap on Debian 10 and I get the same load times as PhieF which for me are acceptable.

**Comment by ChameleonScales** [comment]
3 sec is still 3 times longer than opening the same text through the file browser app.
But 9 times longer means it's not about the server being slow. I don't know what can cause that though.

**Comment by PhieF** [comment]
Because it's not only loading text, it's extracting a zip on a php server and with nextcloud framework creating a file takes a lot of time, on my server even deleting a file takes more than 1 second, so... 

But I've improved things by sending text and metadata as soon as we get them and extracting the note later, here is the result on a bad intel atom based server 

https://crazynoisybizarre.town/system/media_attachments/files/002/371/581/original/00dd470c0139a5e7.mp4?1566114968

**Comment by tgp1994** [comment]
I'd like to report that I'm also seeing issues, when it comes to opening (extract) and DELETEing notes. By watching Firefox's Network page in the dev console, DELETE operations can take anywhere from 5 to 30 seconds to complete. Open (extract) operations take about 20 seconds each. Hopefully there may be ways to optimise performance here? Perhaps a new storage solution may be needed?

**Comment by PhieF** [comment]
Is it on your own server ?

**Comment by tgp1994** [comment]
Yes, locally hosted. I run it in an Ubuntu Hyper-V VM. I also noticed that creating a new note takes a long time to open the note UI.

**Comment by PhieF** [comment]
What about normal nextcloud, can you check how long it takes to delete a file ?

**Comment by tgp1994** [comment]
I tried deleting a file from the webui, and that took about 2-3 seconds tops to respond, about what I'd expect from my installation.

**Comment by PhieF** [comment]
The main problem I guess is that carnet needs to manipulate multiple files, 2-3 is already long so if it needs to manipulate 2 files, then, 4-6 seconds... I will continue to try to make that faster but believe me, nextcloud is slow as hell :D 
notes that take 20 seconds to extract, to they have media or are they only text notes ?

**Comment by tgp1994** [comment]
My notes only consist of text. Thank you for your continued work on this, @PhieF! I'd like to dig into it sometime too and see if I can provide any sort of valuable feedback/info.

**Comment by PhieF** [comment]
Have you considered some nextcloud optimizations like redis, php fpm ? are you use sqlite ? 


You're welcome

**Comment by tgp1994** [comment]
Before writing this comment, I went to my security & config warnings page to double check, and it turns out that I have some db tables needing indexes/conversions/etc. I've now done everything that the page has requested so it shows a green check mark, and I've also installed some system updates that were waiting and rebooted the system. Afterwards I was removing some apparently corrupted notes (showed nothing in them, and got 500 Server errors when deleting) which didn't take as long, although between the POST to action and DELETE commands, it was still taking about 15 seconds for each note.

I did spend quite some time on getting the server optimised. Cron jobs are setup, the system isn't normally loaded with anything else, APCu local memcache is enabled, andI'm using a local mysql db. I don't remember if I've spent much time with FPM and OPCache tuning, I'll try to take a look at that soon and see if it helps anything.

**Comment by gbakeman** [comment]
I'd just like to add here for other people, I was able to increase performance by tweaking how my EXT4 filesystem is mounted. A lot of disk IO was spent in the journal. By changing my root [partition mount options](https://ext4.wiki.kernel.org/index.php/Ext3_Data=Ordered_vs_Data=Writeback_mode) to read `data=writeback`, and making sure I run `tune2fs -o journal_data_writeback` before remounting, I seem to cut load times by about 50%. Instead of the disk IO bottleneck being due to journaling, it's now just due to mysql. There may be some more optimizations I can run to help this. Note: Read the link above for a comparison of pros and cons by doing this. You'll get more performance, but if your PC loses power suddenly in the middle of writing data, there could be some corruption. I have a UPS and weekly backups so I'm not so concerned about this.

I don't know if maybe Carnet runs a lot of database queries when deleting a note? Perhaps these could be reduced if possible?
